packages/devtools_app/lib/src/routing.dart

```diff
@@ -0,0 +1,136 @@
+    return SynchronousFuture<DevToolsRouteConfiguration>(configuration);
+  }
+
+  @override
+  RouteInformation restoreRouteInformation(
+      DevToolsRouteConfiguration configuration) {
+    final path = '/${configuration.page ?? ''}';
+    final params =
+        (configuration.args?.length ?? 0) != 0 ? configuration.args : null;
+    params?.removeWhere((key, value) => value == null);
+    return RouteInformation(
+        location: Uri(path: path, queryParameters: params).toString());
+  }
+}
+
+class DevToolsRouterDelegate extends RouterDelegate<DevToolsRouteConfiguration>
+    with
+        ChangeNotifier,
+        PopNavigatorRouterDelegateMixin<DevToolsRouteConfiguration> {
+  DevToolsRouterDelegate(this._getPage, [GlobalKey<NavigatorState> key])
+      : navigatorKey = key ?? GlobalKey<NavigatorState>();
+
+  static DevToolsRouterDelegate of(BuildContext context) =>
+      Router.of(context).routerDelegate as DevToolsRouterDelegate;
+
+  @override
+  final GlobalKey<NavigatorState> navigatorKey;
+  final Page Function(BuildContext, String, Map<String, String>) _getPage;
+  final routes = ListQueue<DevToolsRouteConfiguration>();
+
+  @override
+  DevToolsRouteConfiguration get currentConfiguration =>
+      routes.isEmpty ? null : routes.last;
+
+  @override
+  Widget build(BuildContext context) {
+    final routeConfig = currentConfiguration;
+    final page = routeConfig?.page;
+    final args = routeConfig?.args ?? {};
+
+    return Navigator(
+      key: navigatorKey,
+      pages: [_getPage(context, page, args)],
+      onPopPage: (route, result) => popPage(),
+    );
+  }
+
+  bool popPage() {
+    if (routes.length <= 1) {
+      return false;
+    }
+
+    routes.removeLast();
+    notifyListeners();
+    return true;
+  }
+
+  /// Pushes a new page, optionally updating arguments.
+  ///
+  /// Existing arguments (for example &uri=) will be preserved unless
+  /// overwritten by [updateArgs].
+  void pushPageIfNotCurrent(String page, [Map<String, String> updateArgs]) {
+    final pageChanged = page != currentConfiguration.page;
+    final argsChanged = !mapEquals(
+      {...currentConfiguration.args, ...?updateArgs},
+      currentConfiguration.args,
+    );
+    if (!pageChanged && !argsChanged) {
+      return;
+    }
+
+    routes.add(DevToolsRouteConfiguration(
+        page, {...currentConfiguration.args, ...?updateArgs}));
+
+    notifyListeners();
+  }
+
+  @override
+  Future<void> setNewRoutePath(DevToolsRouteConfiguration configuration) {
+    routes.add(configuration);
+    return SynchronousFuture<void>(null);
+  }
+
+  /// Updates arguments for the current page.
+  ///
+  /// Existing arguments (for example &uri=) will be preserved unless
+  /// overwritten by [updateArgs].
+  void updateArgsIfNotCurrent(Map<String, String> updateArgs) {
+    final argsChanged = !mapEquals(
+      {...currentConfiguration.args, ...?updateArgs},
+      currentConfiguration.args,
+    );
```

Comment:
let's put this in a helper since it is used in multiple places `_argsChanged({@required Map<String, String> current, @required Map<String, String> updated})`

Reply:
Done - though since the first arg was always the current args (and calling it was just as verbose as the original code with named args), I made the helper just always compare against the current args.